markovflow/sde/sde.py

```diff
@@ -0,0 +1,165 @@
+        """
+        Initialize the Ornstein-Uhlenbeck SDE.
+
+        :param decay: λ, a tensor with shape ``(1, 1)``.
+        :param q: spectral density of the Brownian motion ``(state_dim, state_dim)``.
+        """
+        super(OrnsteinUhlenbeckSDE, self).__init__()
+        self.decay = decay
+        self.q = tf.cast(q, dtype=decay.dtype)
+        self.state_dim = q.shape[0]
+
+    def drift(self, x: tf.Tensor, t: tf.Tensor) -> tf.Tensor:
+        """
+        Drift of the Ornstein-Uhlenbeck process
+        ..math:: f(x(t), t) = -λ x(t)
+
+        :param x: state at `t` i.e. `x(t)` with shape ``(n_batch, state_dim)``.
+        :param t: time `t` with shape ``(n_batch, 1)``.
+
+        :return: Drift value i.e. `f(x(t), t)` with shape ``(n_batch, state_dim)``.
+        """
+        assert x.shape[-1] == self.state_dim
+        return -self.decay * x
+
+    def diffusion(self, x: tf.Tensor, t: tf.Tensor) -> tf.Tensor:
+        """
+        Diffusion of the Ornstein-Uhlenbeck process
+        ..math:: l(x(t), t) = sqrt(q)
+
+        :param x: state at `t` i.e. `x(t)` with shape ``(n_batch, state_dim)``.
+        :param t: time `t` with shape ``(n_batch, 1)``.
+
+        :return: Diffusion value i.e. `l(x(t), t)` with shape ``(n_batch, state_dim, state_dim)``.
+        """
+        assert x.shape[-1] == self.state_dim
+        return tf.linalg.cholesky(self.q)
+
+
+class DoubleWellSDE(SDE):
+    """
+    Double-Well SDE represented by
+
+    ..math:: dx(t) = f(x(t)) dt + dB(t),
+
+    where f(x(t)) = 4 x(t) (1 - x(t)^2) and the spectral density of the Brownian motion is specified by q.
+    """
+    def __init__(self, q: tf.Tensor = tf.ones((1, 1))):
+        """
+        Initialize the Double-Well SDE.
+
+        :param q: spectral density of the Brownian motion ``(state_dim, state_dim)``.
+        """
+        super(DoubleWellSDE, self).__init__()
+        self.q = q
+        self.state_dim = q.shape[0]
+
+    def drift(self, x: tf.Tensor, t: tf.Tensor) -> tf.Tensor:
+        """
+        Drift of the double-well process
+        ..math:: f(x(t), t) = 4 x(t) (1 - x(t)^2)
+
+        :param x: state at `t` i.e. `x(t)` with shape ``(n_batch, state_dim)``.
+        :param t: time `t` with shape ``(n_batch, 1)``.
+
+        :return: Drift value i.e. `f(x(t), t)` with shape ``(n_batch, state_dim)``.
+        """
+        assert x.shape[-1] == self.state_dim
+        return 4. * x * (1. - tf.square(x))
+
+    def diffusion(self, x: tf.Tensor, t: tf.Tensor) -> tf.Tensor:
+        """
+        Diffusion of the double-well process
+        ..math:: l(x(t), t) = sqrt(q)
+
+        :param x: state at `t` i.e. `x(t)` with shape ``(n_batch, state_dim)``.
+        :param t: time `t` with shape ``(n_batch, 1)``.
+
+        :return: Diffusion value i.e. `l(x(t), t)` with shape ``(n_batch, state_dim, state_dim)``.
+        """
+        assert x.shape[-1] == self.state_dim
+        self.q = tf.cast(self.q, dtype=x.dtype)
```

Comment:
I get to know the dtype of x here and need to made q of the same type otherwise in the simulation method say EM it will throw an error. Should it be casted outside or trust that q is always passed in the correct dtype?